backend/src/email/middlewares/email-template.middleware.ts

```diff
@@ -255,16 +262,103 @@ const deleteCsvErrorHandler = async (
 ): Promise<Response | void> => {
   try {
     const { campaignId } = req.params
-    await TemplateService.deleteS3TempKeys(+campaignId)
+    await UploadService.deleteS3TempKeys(+campaignId)
     res.sendStatus(200)
   } catch (e) {
     next(e)
   }
 }
 
+// TODO: refactor this handler with uploadCompleteHandler to share the same function
+/**
+ * Downloads the file from s3 and checks that its columns match the attributes provided in the template.
+ * If a template has not yet been uploaded, do not write to the message logs, but prompt the user to upload a template first.
+ * If the template and csv do not match, prompt the user to upload a new file.
+ * @param req
+ * @param res
+ * @param next
+ */
+const uploadProtectedCompleteHandler = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+): Promise<Response | void> => {
+  try {
+    const { campaignId } = req.params
+
+    if (!(await EmailTemplateService.getFilledTemplate(+campaignId))) {
+      throw new Error('Template does not exist, please create a template')
+    }
+
+    // extract s3Key from transactionId
+    const { transaction_id: transactionId, filename } = req.body
+    const { s3Key } = UploadService.extractParamsFromJwt(transactionId) as {
+      s3Key: string
+      uploadId: string
+    }
+
+    // - download from s3
+    const s3Client = new S3Client()
+    const fileContent = await s3Client.getCsvFile(s3Key)
+
+    const records = UploadService.getProtectedRecordsFromCsv(
+      +campaignId,
+      fileContent
+    )
+
+    if (EmailTemplateService.hasInvalidEmailRecipient(records)) {
```

Comment:
It seems alittle too late to check the emails here, because the file would have been encrypted and uploaded by then. Could we move this check to the frontend?